onnxruntime/core/optimizer/matmul_scale_fusion.cc

```diff
@@ -0,0 +1,282 @@
+
+  if (graph_utils::IsSupportedOptypeVersionAndDomain(scale_node, "Div", {7})) {
+    // (x / scale_reciprocal)
+    const auto div_inputs = scale_node.InputDefs();
+    ORT_ENFORCE(div_inputs.size() == 2);
+
+    const int scale_reciprocal_arg_index = 1;
+    const NodeArg& scale_reciprocal_node_arg = *div_inputs[scale_reciprocal_arg_index];
+
+    if (is_excluded_initializer(scale_reciprocal_node_arg)) return {};
+
+    const auto divisor = GetScalarConstantInitializer(graph, scale_reciprocal_node_arg);
+
+    if (!divisor.has_value()) return {};
+
+    return {std::make_pair(1.0f / divisor.value(), scale_reciprocal_arg_index)};
+  }
+
+  if (graph_utils::IsSupportedOptypeVersionAndDomain(scale_node, "Mul", {7})) {
+    // (x * scale) or (scale * x)
+    const auto mul_inputs = scale_node.InputDefs();
+    ORT_ENFORCE(mul_inputs.size() == 2);
+
+    for (int scale_arg_index = 0; scale_arg_index < 2; ++scale_arg_index) {
+      const NodeArg& scale_node_arg = *mul_inputs[scale_arg_index];
+
+      if (is_excluded_initializer(scale_node_arg)) continue;
+
+      const auto multiplier = GetScalarConstantInitializer(graph, scale_node_arg);
+
+      if (!multiplier.has_value()) continue;
+
+      return {std::make_pair(multiplier.value(), scale_arg_index)};
+    }
+
+    return {};
+  }
+
+  return {};
+}
+
+struct ScaleMergeInfo {
+  // the edge from the base node to the original node
+  Node::EdgeConstIterator node_to_merge_edge;
+  // the scale of the original node
+  float scale;
+  // the index of the input or output def on the original node
+  // this def is moved to the fused node
+  // for a leading scale (scale -> MatMul), it will be the unscaled input
+  // for a trailing scale (MatMul -> scale), it will be the scaled output
+  int node_to_merge_def_index;
+  // the index of the input or output def on the fused node
+  int fused_node_def_index;
+};
+
+std::vector<ScaleMergeInfo> GetInputNodeMerges(
+    Graph& graph, Node& node,
+    const std::unordered_set<std::string>& excluded_initializer_names) {
+  std::vector<ScaleMergeInfo> input_node_merges{};
+  for (auto input_edge = node.InputEdgesBegin(); input_edge != node.InputEdgesEnd(); ++input_edge) {
+    const Node& input_node = input_edge->GetNode();
+
+    if (input_node.GetExecutionProviderType() != node.GetExecutionProviderType()) continue;
+    const auto scale_and_index = GetScaleFromNode(graph, input_node, excluded_initializer_names);
+    if (!scale_and_index.has_value()) continue;
+
+    // assume scale nodes have 2 input defs, so to_scale_index == 1 - scale_index
+    ORT_ENFORCE(input_node.InputDefs().size() == 2 && scale_and_index.value().second < 2);
+    const int to_scale_index = 1 - scale_and_index.value().second;
+
+    input_node_merges.push_back(
+        {input_edge,
+         scale_and_index.value().first,
+         to_scale_index,
+         input_edge->GetDstArgIndex()});
+  }
+  return input_node_merges;
+}
+
+std::vector<ScaleMergeInfo> GetOutputNodeMerges(
```

Comment:
>GetOutputNodeMerges [](start = 28, length = 19)

Can output have multiple nodes to merge? 

if you have two nodes consumes output Y, one div by 8, another one div by 16. We actually should apply fusion here.

Reply:
that's possible. then it's transforming the original subgraph into multiple TransposeScaleMatMul nodes possibly. maybe a future enhancement for another PR? it adds some complexity and is not required for the case we're targeting now.

---
In reply to: [464695609](https://github.com/microsoft/onnxruntime/pull/4669#discussion_r464695609) [](ancestors = 464695609)

Reply:
Sorry, I had a typo in my comments above. 
I meant to say "We shouldn't apply fusion here". 

Fusing them into 2 different op will incur duplicated computation. We should skip fusion in this case. 

---
In reply to: [464700432](https://github.com/microsoft/onnxruntime/pull/4669#discussion_r464700432) [](ancestors = 464700432,464695609)

Reply:
ah, got it. the condition `optimizer_utils::CheckOutputEdges(graph, node, 1)` should ensure there is exactly one output edge (and no graph output) on the node.

---
In reply to: [464720046](https://github.com/microsoft/onnxruntime/pull/4669#discussion_r464720046) [](ancestors = 464720046,464700432,464695609)